### ISSUE TITLE
Ruff wrongly flags valid mypy forward references in TYPE_CHECKING as not defined

### ISSUE THREAD

**Issue by ezyang** [astral-sh/ruff#9753]
```
from typing import TYPE_CHECKING

class A:
    if TYPE_CHECKING:
        B = A
```

fails with

```
a.py:5:13: F821 Undefined name `A`
```

but this is fine by mypy.

**Comment by charliermarsh** [comment]
Interesting... In my cursory testing, Pyright doesn't respect this, so I'm not certain that we should? \cc @erictraut in case you have thoughts.

**Comment by erictraut** [comment]
Mypy's failure to detect this has nothing to do with `TYPE_CHECKING`. It doesn't detect invalid forward references even if you were to remove the `TYPE_CHECKING` conditional:

```python
class A:
    B = A # No error from mypy
```

I think pyright and ruff are correct in flagging this as an error. If mypy were to add this missing functionality, it would likewise report an error here.

In general, I recommend against using `TYPE_CHECKING` within your code. It's effectively a "lie" to static analysis tools and type checkers, and that can mask errors and make your code more fragile. 

If your intent here is to define a type alias `B` within the scope of class `A`, you can do so without the use of `TYPE_CHECKING` as follows...

If you're using Python 3.12:
```python
class A:
    type B = A
```

If you're using an older version of Python:
```python
from typing import TypeAlias
class A:
    B: TypeAlias = "A"
```

**Comment by AlexWaygood** [comment]
I agree with Eric -- this isn't a feature of mypy; OP's snippet onlly works with mypy due to a feature being missing from mypy

**Comment by charliermarsh** [comment]
I'm gonna close this for now based on the above feedback from @erictraut and @AlexWaygood. Sorry for any inconvenience @ezyang.
